File: tests/e2e/dotnet/copy_ops.cs
// Copies a real in-cone input into an overlay directory with File.Copy, then
// reads the copy back and lists the directory. File.Copy maps to the Win32
// CopyFile / CopyFileEx kernel-copy path (the runtime opens the source and
// creates the destination itself), so the overlay must intercept those opens
// and redirect the write into the backing store - a leak would land out.txt on
// the real execroot. This covers the CopyFile family (`File.Copy`) separately
// from fs_ops.cs's File.Move.
//
// Run as: copy_ops.exe <execroot>
// Emits:
//   READ=<content>   content read back from the copied file
//   LIST=<entries>   directory listing of the overlay directory
using System;
using System.IO;
using System.Linq;

internal static class Program {
    private static int Main(string[] args) {
        if (args.Length < 1) {
            Console.Error.WriteLine("usage: copy_ops.exe <execroot>");
            return 2;
        }
        string ws = args[0];

        string src = Path.Combine(ws, "in.txt");  // real, in-cone input (allowed read)
        File.WriteAllText(src, "OVNETCOPY");

        string d = Path.Combine(ws, "wd");
        Directory.CreateDirectory(d);
        string dst = Path.Combine(d, "out.txt");
        File.Copy(src, dst);

        string read = File.ReadAllText(dst);
        string listing = string.Join(
            ",",
            Directory.GetFileSystemEntries(d)
                .Select(Path.GetFileName)
                .OrderBy(name => name, StringComparer.Ordinal));

        Console.Write("READ=" + read + " LIST=" + listing);
        return 0;
    }
}

File: tests/e2e/python/scripts/copy_ops.py
# Copies a real in-cone input into an overlay directory with shutil.copy, then
# reads the copy back and lists the directory - the overlay must intercept the
# source read and redirect the destination write into the backing store (a leak
# would land out.txt on the real execroot).
#
# shutil.copy is the stdlib file-copy tool; on CPython it
# opens the source, creates the destination, and copies the bytes (plus a
# copymode chmod on the destination), so it drives read-after-write on the copy
# and an enumeration splice on the overlay directory in one shot.
#
# Run as: python copy_ops.py <execroot>
# Emits:
#   READ=<content>     content read back from the copied file
#   LIST=<entries>     os.scandir listing of the overlay directory
import os
import shutil
import sys


def main():
    if len(sys.argv) < 2:
        sys.stderr.write("usage: copy_ops.py <execroot>\n")
        return 2
    ws = sys.argv[1]

    src = os.path.join(ws, "in.txt")  # real, in-cone input (allowed read)
    with open(src, "w", encoding="utf-8") as f:
        f.write("OVPYCOPY")

    d = os.path.join(ws, "wd")
    os.makedirs(d, exist_ok=True)
    dst = os.path.join(d, "out.txt")
    shutil.copy(src, dst)

    with open(dst, "r", encoding="utf-8") as f:
        read = f.read()
    with os.scandir(d) as it:
        listing = ",".join(sorted(e.name for e in it))

    sys.stdout.write("READ=" + read + " LIST=" + listing)
    return 0


if __name__ == "__main__":
    sys.exit(main())
